sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveDataCache.java

```diff
@@ -0,0 +1,218 @@
+     * {@code TransferInfo} LiveData
+     */
+    @MainThread
+    public TransferIdInfoLiveData.LiveDataPair create(long transferId, Context context) {
+        this.expunge();
+        final TransferIdInfoLiveData.Result result = TransferIdInfoLiveData.create(context);
+        final TransferIdInfoLiveData.LiveDataPair liveDataPair = result.getLiveDataPair();
+        this.map.put(transferId,
+            new TransferInfoLiveDataWeakReference(liveDataPair.getTransferInfoLiveData(),
+                liveDataPair.getTransferIdOrErrorLiveData(),
+                result.getTransferFlags(),
+                transferId,
+                this.queue));
+        return liveDataPair;
+    }
+
+    /**
+     * Check whether the TransferIdOrError LiveData and the TransferInfo LiveData that is identified
+     * by the given {@code transferId} key already exists in the cache, if it exists then return
+     * the two LiveData in a {@Link TransferIdInfoLiveData#Pair}. If it does not exists then create,
+     * store and return them, see {@link TransferIdInfoLiveDataCache#create(long, Context)}
+     * for more details.
+     *
+     * @param transferId the transferId to use as the key to identify the LiveData instances.
+     * @param context the context
+     * @return the pair composing created TransferIdOrError LiveData and associated TransferInfo LiveData
+     */
+    @MainThread
+    public TransferIdInfoLiveData.LiveDataPair getOrCreate(long transferId, Context context) {
+        this.expunge();
+        final TransferInfoLiveDataWeakReference ref = this.map.get(transferId);
+        if (ref != null) {
+            final LiveData<TransferInfo> transferInfoLiveData = ref.get();
+            if (transferInfoLiveData != null) {
+                return new TransferIdInfoLiveData.LiveDataPair(ref.getTransferIdLiveData(),
+                    transferInfoLiveData);
+            }
+        }
+        return create(transferId, context);
+    }
+
+    /**
+     * Get the Flag object that the TransferInfo LiveData source (that backs TransferInfo LiveData
+     * in the weak reference) check for any flag set by TransferClient methods on the transfer.
+     *
+     * @param transferId the transferId identifies the {@link TransferFlags} for the transfer
+     * @return the {@link TransferFlags}
+     */
+    @MainThread
+    public TransferIdInfoLiveData.TransferFlags getTransferFlags(long transferId) {
+        this.expunge();
+        final TransferInfoLiveDataWeakReference ref = this.map.get(transferId);
+        if (ref != null) {
+            return ref.getTransferFlags();
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Expunges stale entries (containing already GCed {@code TransferInfo} LiveData) from the map.
+     */
+    @SuppressWarnings("unchecked")
+    private void expunge() {
+        int cnt = 0;
+        for (Reference<?> staledReference; (staledReference = this.queue.poll()) != null;) {
```

Comment:
Is this not just the same as a while loop? I'm fine with it but seeing this for-loop syntax is more confusing to me.

Reply:
I see, yea `while` is more cleaner/readable, will switch to it.